include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -250,118 +257,341 @@ __parallel_for(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&
                     if (__local_idx == 0)
                     {
-                        //final reduction
-                        if (__n_groups == 1)
-                            __brick_reduce.apply_init(__init, __result);
-
-                        __temp_acc[__offset_1 + __item_id.get_group(0)] = __result;
+                        __reduce_pattern.apply_init(__init, __result);
+                        __res_acc[0] = __result;
                     }
                 });
         });
-        if (__is_first)
-            __is_first = false;
-        ::std::swap(__offset_1, __offset_2);
-        __n = __n_groups;
-        __n_items = (__n - 1) / __iters_per_work_item + 1;
-        __n_groups = (__n - 1) / __size_per_work_group + 1;
-    } while (__n > 1);
-
-    return __future(__reduce_event, sycl::buffer(__temp, sycl::id<1>(__offset_2), sycl::range<1>(1)));
+
+        return __future(__reduce_event, __res);
+    }
+}; // struct __parallel_transform_reduce_small_submitter
+
+// General parallel_transform_reduce - uses a tree-based reduction
+template <typename _Tp, typename _IsGPU>
+struct __parallel_transform_reduce_submitter
+{
+    template <typename... _Name>
+    using _KernelName = __reduce_kernel<_IsGPU{}, _Name...>;
+
+    template <typename _ExecutionPolicy, typename _Size, typename _ReduceOp, typename _TransformOp1,
+              typename _TransformOp2, typename _InitType,
+              oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0,
+              typename... _Ranges>
+    static auto
+    submit(_ExecutionPolicy&& __exec, _Size __n, ::std::uint16_t __work_group_size, _ReduceOp __reduce_op,
+           _TransformOp1 __transform_op1, _TransformOp2 __transform_op2, _InitType __init, _Ranges&&... __rngs)
+    {
+        using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
+        using _CustomName = typename _Policy::kernel_name;
+        using _ReduceKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
+            _KernelName, _CustomName, _ReduceOp, _TransformOp1, _TransformOp2, _Ranges...>;
+
+#if _ONEDPL_COMPILE_KERNEL
+        auto __kernel = __internal::__kernel_compiler<_ReduceKernel>::__compile(
+            ::std::forward<_ExecutionPolicy>(__exec));
+        __work_group_size =
+            ::std::min(__work_group_size, (::std::uint16_t)oneapi::dpl::__internal::__kernel_work_group_size(
+                                              ::std::forward<_ExecutionPolicy>(__exec), __kernel));
+#endif
+
+        // __iters_per_work_item shows number of elements to reduce on global memory
+        // __work_group_size shows number of elements to reduce on local memory
+        _Size __iters_per_work_item = 1;
+        if (_IsGPU{})
+        {
+            // empirically tested launch configuration
+            __iters_per_work_item = 32;
+            __work_group_size = ::std::min(__work_group_size, (::std::uint16_t)256);
```

Comment:
`(::std::uint16_t)256`
do we really need it?

Reply:
Yes, since `__work_group_size` is a `uint16_t` and `std::min()` expects two parameters of the same type. Also, there is no unsigned short literal to use.